src/mongo/client/dbclient.cpp

```diff
@@ -1066,15 +1070,175 @@ namespace mongo {
+                                                     "todb" << todb);
+
+            BSONObj saslCommandPrefix = saslFirstCommandPrefix;
+            BSONObj inputObj = BSON(saslCommandPayloadFieldName << "");
+            bool isServerDone = false;
+
+            // send copydbsaslstart, then continue to send copydb until we are done.
+            while (!session->isDone()) {
+                std::string payload;
+                BSONType type;
+
+                Status status = saslExtractPayload(inputObj, &payload, &type);
+                if (!status.isOK()) {
+                    throw DBException( str::stream() << "sasl session failure: " << status.reason(), 14355 );
+                }
+
+                std::string responsePayload;
+                status = session->step(payload, &responsePayload);
+                if (!status.isOK()) {
+                    throw DBException( str::stream() << "sasl session failure: " << status.reason(), 14356 );
+                }
+
+                // build command to send to server
+                BSONObjBuilder commandBuilder;
+                commandBuilder.appendElements(saslCommandPrefix);
+                commandBuilder.appendBinData(saslCommandPayloadFieldName,
+                                             int(responsePayload.size()),
+                                             BinDataGeneral,
+                                             responsePayload.c_str());
+                BSONElement conversationId = inputObj[saslCommandConversationIdFieldName];
+                if (!conversationId.eoo()) {
+                    commandBuilder.append(conversationId);
+                }
+
+                BSONObj command = commandBuilder.obj();
+                bool ok = runCommand("admin", command, inputObj);
+
+                ErrorCodes::Error code =
+                    ErrorCodes::fromInt(inputObj[saslCommandCodeFieldName].numberInt());
+
+                if (!ok || code != ErrorCodes::OK) {
+                    if (code == ErrorCodes::OK)
+                        code = ErrorCodes::UnknownError;
+                    // TODO handle
+                    // attempt to give a sane error message
+                    if ( inputObj.hasField("errmsg") ) {
+                        BSONElement e = inputObj.getField("errmsg");
+                        if (e.type() == String &&
+                            strstr( e.valuestr(), "no such cmd: saslStart")) {
+                            throw UserException( 13423,
+                                                 "fromhost doesn't support SCRAM-SHA-1, use MONGODB-CR");
+                        }
+                    }
+                    throw OperationException( inputObj );
+                }
+
+                isServerDone = inputObj[saslCommandDoneFieldName].trueValue();
+                saslCommandPrefix = saslFollowupCommandPrefix;
+            }
+            if (!isServerDone) {
+                assert(false); // TODO throw something better
```

Comment:
Well it is an invariant on the SASL mechanism used (SCRAM-SHA-1) in this case and the client and server implementations. The client "should" never finish before the server but it could theoretically happen if some version of the server considers SCRAM-SHA-1 taking more iterations then client.